File: tests/fixtures/core-sources.ts
import type {CoreEvidenceLink,CoreSourceCatalog} from "../../packages/contracts/src/spatial/core";
import {identityFixture,ref} from "./core-identity";

export const revisionRef=(id:string,namespace:string,revision=1)=>({ref:ref(id,namespace),revision});
export function evidence(id:string,target="B1",part="p1"):CoreEvidenceLink {
  return {ref:ref(id,"evidence_link") as CoreEvidenceLink['ref'],revision:1,target:ref(target),part:revisionRef(part,"source_part") as CoreEvidenceLink['part'],purpose:"context",state:"active",inheritance:{kind:"direct"}};
}
export function sourceFixture():CoreSourceCatalog {
  return {
    datasets:[{ref:{namespace:"dataset",id:"dataset-1"},revision:1,label:"Authored test sources",classification:"synthetic",attribution:"Synthetic fixture",license:null,access:"operator"}],
    assets:[{ref:{namespace:"asset",id:"a1"},revision:1,kind:"original",mediaType:"application/pdf",sha256:'a'.repeat(64),bytes:123,
      storage:{state:"registered",blobRef:{namespace:"source_blob",id:"source-original"}},integrity:"metadata_only",access:"operator",retention:{policy:"preserve_original",legalHold:false},parentAssets:[]}],
    sources:[{ref:{namespace:"source_revision",id:"s1"},revision:1,family:{namespace:"source_family",id:"family-1"},familyOrdinal:3,label:"Private filename",profile:"plan-pdf-v1",method:"synthetic",dataset:{ref:{namespace:"dataset",id:"dataset-1"},revision:1},assets:[{ref:{namespace:"asset",id:"a1"},revision:1}],workflows:[ref("case-1","case")],access:"operator"}],
    parts:[{ref:{namespace:"source_part",id:"p1"},revision:1,source:{ref:{namespace:"source_revision",id:"s1"},revision:1},asset:{ref:{namespace:"asset",id:"a1"},revision:1},locators:[{kind:"page",page:2,region:{x:0.1,y:0.2,width:0.3,height:0.4,unit:"normalized"}},{kind:"verbatim",locator:"Page 2 · retained original locator"}],access:"operator"}],
    links:[evidence("building","B1"),evidence("other-building","B2"),{...evidence("inherited-unit","U1"),inheritance:{kind:"inherited",parentLink:{ref:{namespace:"evidence_link",id:"building"},revision:1},via:["duplex-parent"]}}],
  };
}
export interface SourceConformanceCase {id:string;catalog:unknown;identity:unknown;valid:boolean;code?:string}
export function sourceCases():SourceConformanceCase[] {
  const cases:SourceConformanceCase[]=[{id:"shared-original-and-inherited-context",catalog:sourceFixture(),identity:identityFixture(),valid:true},{id:"zero-document-catalog",catalog:{datasets:[],assets:[],sources:[],parts:[],links:[]},identity:identityFixture(),valid:true}];
  const geometry:any=JSON.parse(JSON.stringify(sourceFixture()));geometry.assets[0].mediaType="application/geo+json; charset=utf-8";geometry.sources[0].profile="geojson";geometry.parts[0].locators=[{kind:"feature",featureId:"0000123"},{kind:"json_pointer",pointer:"/features/0"}];geometry.links[0].purpose="geometry";
  cases.push({id:"geometry-evidence-with-zero-pdfs",catalog:geometry,identity:identityFixture(),valid:true});
  const archived:any=JSON.parse(JSON.stringify(sourceFixture()));
  archived.linkHistory=[JSON.parse(JSON.stringify(archived.links[2])),JSON.parse(JSON.stringify(archived.links[0]))];
  archived.links[2].state="unlinked";archived.links[2].revision=2;
  archived.links[0].state="unlinked";archived.links[0].revision=2;
  cases.push({id:"child-and-parent-unlinked-with-exact-history",catalog:archived,identity:identityFixture(),valid:true});
  const historyChange=(id:string,code:string,mutate:(catalog:any)=>void)=>{
    const catalog=JSON.parse(JSON.stringify(archived));mutate(catalog);
    cases.push({id,catalog,identity:identityFixture(),valid:false,code});
  };
  historyChange("archived-reference-needs-actual-record","STALE_REFERENCE",c=>c.linkHistory=c.linkHistory.filter((l:any)=>l.ref.id!=="building"));
  historyChange("active-child-cannot-inherit-historical-parent","STALE_REFERENCE",c=>c.links[2].state="active");
  historyChange("duplicate-evidence-history","DUPLICATE_RECORD",c=>c.linkHistory.push(c.linkHistory[0]));
  historyChange("history-needs-current-identity","MISSING_REFERENCE",c=>c.linkHistory[0].ref.id="orphan-history");
  historyChange("future-evidence-history","HISTORY_REVISION",c=>c.linkHistory[0].revision=3);
  historyChange("current-revision-is-not-history","HISTORY_REVISION",c=>c.linkHistory[0].revision=2);
  const change=(id:string,code:string,mutate:(catalog:any,identity:any)=>void)=>{const catalog=JSON.parse(JSON.stringify(sourceFixture())),identity=JSON.parse(JSON.stringify(identityFixture()));mutate(catalog,identity);cases.push({id,catalog,identity,valid:false,code});};
  change("duplicate-asset","DUPLICATE_RECORD",c=>c.assets.push(c.assets[0]));
  change("unknown-source-ordinal-not-zero","INVALID_CONTRACT",c=>c.sources[0].familyOrdinal=0);
  change("missing-dataset","MISSING_REFERENCE",c=>c.sources[0].dataset.ref.id="missing");
  change("stale-original-asset","STALE_REFERENCE",c=>c.sources[0].assets[0].revision=2);
  change("original-marked-disposable","ORIGINAL_RETENTION",c=>c.assets[0].retention.policy="managed_derivative");
  change("registered-original-without-hash","ASSET_METADATA",c=>c.assets[0].sha256=null);
  change("unqualified-integrity-claim","INVALID_CONTRACT",c=>c.assets[0].integrity="verified");
  change("asset-parent-cycle","ASSET_CYCLE",c=>{c.assets[0].kind="derived";c.assets[0].parentAssets=[revisionRef("a1","asset")];});
  change("stale-source-part","STALE_REFERENCE",c=>c.parts[0].source.revision=2);
  change("exact-part-without-original-asset","PART_ASSET",c=>c.parts[0].asset=null);
  change("reverse-row-range","LOCATOR_RANGE",c=>c.parts[0].locators=[{kind:"rows",range:{start:3,end:2}}]);
  change("crop-outside-original-page","LOCATOR_REGION",c=>c.parts[0].locators[0].region.width=0.95);
  change("unsafe-json-pointer-escape","INVALID_CONTRACT",c=>c.parts[0].locators=[{kind:"json_pointer",pointer:"/bad~2escape"}]);
  change("duplicate-locator","DUPLICATE_REFERENCE",c=>c.parts[0].locators.push(c.parts[0].locators[0]));
  change("unknown-evidence-target","MISSING_TARGET",c=>c.links[0].target=ref("missing"));
  change("stale-evidence-part","STALE_REFERENCE",c=>c.links[0].part.revision=2);
  change("verbatim-not-exact-geometry","LOCATOR_NOT_QUALIFIED",c=>{c.parts[0].locators=[{kind:"verbatim",locator:"somewhere"}];c.links[0].purpose="geometry";});
  change("root-pointer-not-exact-feature","LOCATOR_NOT_QUALIFIED",c=>{c.parts[0].locators=[{kind:"json_pointer",pointer:""}];c.links[0].purpose="geometry";});
  change("missing-inherited-parent","MISSING_REFERENCE",c=>c.links[2].inheritance.parentLink.ref.id="missing");
  change("stale-inherited-parent","STALE_REFERENCE",c=>c.links[2].inheritance.parentLink.revision=3);
  change("inherited-geometry-is-not-unit-evidence","INHERITANCE_PURPOSE",c=>c.links[2].purpose="geometry");
  change("inheritance-via-service-not-containment","INHERITANCE_PATH",c=>{c.links[2].target=ref("pipe");c.links[2].inheritance.via=["pipe-service"];});
  change("wrong-parent-building","INHERITANCE_PATH",c=>c.links[2].inheritance.parentLink.ref.id="other-building");
  change("inheritance-from-unlinked-parent","INHERITANCE_PARENT",c=>c.links[0].state="unlinked");
  change("duplicate-family-ordinal","SOURCE_ORDINAL",c=>c.sources.push({...c.sources[0],ref:{namespace:"source_revision",id:"second-source"}}));
  change("conflicting-shared-blob-metadata","BLOB_METADATA",c=>c.assets.push({...c.assets[0],ref:{namespace:"asset",id:"second-asset"},sha256:'b'.repeat(64)}));
  return cases;
}

File: services/geo/geo/core_sources.py
"""Source/evidence referential semantics from the common generated contract."""
from .core_contract import bundled_schema, fail, index_records, parse_core, ref_key, require_revision, structural_key, version_key
from .core_identity import validate_core_identity_graph


def unique(values, kind):
    if len(set(values)) != len(values):
        fail("DUPLICATE_REFERENCE", "Duplicate " + kind + " association")


def validate_core_source_catalog(value, identity_value):
    catalog = parse_core("source-catalog", value)
    identity = validate_core_identity_graph(identity_value)
    policy = bundled_schema("source-catalog")[0]["x-ulpin-evidence-policy"]
    entities = {ref_key(e["ref"]) for e in identity["entities"]}
    relations = {r["id"]: r for r in identity["relations"]}
    datasets = index_records(catalog["datasets"], "dataset")
    assets = index_records(catalog["assets"], "asset")
    sources = index_records(catalog["sources"], "source")
    parts = index_records(catalog["parts"], "source part")
    links = index_records(catalog["links"], "evidence link")
    history = {}
    for entry in catalog.get("linkHistory", []):
        current = links.get(ref_key(entry["ref"]))
        key = version_key(entry)
        if current is None:
            fail("MISSING_REFERENCE", "Evidence history needs a current association record")
        if entry["revision"] >= current["revision"]:
            fail("HISTORY_REVISION", "Evidence history must precede the current revision")
        if key in history:
            fail("DUPLICATE_RECORD", "Duplicate evidence history revision")
        history[key] = entry
    blobs, parents = {}, {}
    for asset in catalog["assets"]:
        if asset["kind"] == "original" and asset["retention"]["policy"] != "preserve_original":
            fail("ORIGINAL_RETENTION", "Original receipts are not disposable derivatives")
        if asset["storage"]["state"] == "registered":
            if asset["sha256"] is None or asset["bytes"] is None:
                fail("ASSET_METADATA", "Registered asset metadata requires checksum and bytes")
            key = ref_key(asset["storage"]["blobRef"])
            metadata = asset["sha256"], asset["bytes"], asset["mediaType"]
            if key in blobs and blobs[key] != metadata:
                fail("BLOB_METADATA", "One blob has conflicting metadata")
            blobs[key] = metadata
        unique([version_key(link) for link in asset["parentAssets"]], "asset parent")
        parents[ref_key(asset["ref"])] = [ref_key(require_revision(assets, link, "asset parent")["ref"]) for link in asset["parentAssets"]]
    complete = set()
    for root in assets:
        pending, active = [(root, False)], set()
        while pending:
            key, leave = pending.pop()
            if leave:
                active.remove(key); complete.add(key)
                continue
            if key in complete:
                continue
            if key in active:
                fail("ASSET_CYCLE", "Asset ancestry contains a cycle")
            active.add(key)
            pending.append((key, True))
            pending.extend((parent, False) for parent in parents.get(key, []))
    ordinals = []
    for source in catalog["sources"]:
        if source["dataset"] is not None:
            require_revision(datasets, source["dataset"], "dataset")
        for asset in source["assets"]:
            require_revision(assets, asset, "source asset")
        unique([version_key(link) for link in source["assets"]], "source asset")
        unique([ref_key(link) for link in source["workflows"]], "source workflow")
        if source["family"] is not None and source["familyOrdinal"] is not None:
            ordinals.append((ref_key(source["family"]), source["familyOrdinal"]))
    if len(set(ordinals)) != len(ordinals):
        fail("SOURCE_ORDINAL", "A source ordinal refers to multiple immutable receipts")
    for part in catalog["parts"]:
        source = require_revision(sources, part["source"], "source part's source")
        if part["asset"] is not None:
            require_revision(assets, part["asset"], "part asset")
            if version_key(part["asset"]) not in [version_key(link) for link in source["assets"]]:
                fail("PART_ASSET", "Part asset does not belong to its source")
        elif source["assets"] or any(locator["kind"] not in ("model_element", "verbatim") for locator in part["locators"]):
            fail("PART_ASSET", "An exact file part must name its original asset")
        unique([structural_key(locator) for locator in part["locators"]], "source locator")
        for locator in part["locators"]:
            if locator["kind"] in ("rows", "lines") and locator["range"]["end"] < locator["range"]["start"]:
                fail("LOCATOR_RANGE", "Source range is reversed")
            region = locator.get("region") if locator["kind"] in ("page", "image_region") else None
            if region and (region["x"] + region["width"] > 1 + policy["normalizedTolerance"] or region["y"] + region["height"] > 1 + policy["normalizedTolerance"]):
                fail("LOCATOR_REGION", "Source region lies outside the original")
    edge_keys = []
    for link in catalog["links"] + catalog.get("linkHistory", []):
        is_current = links[ref_key(link["ref"])] is link
        if ref_key(link["target"]) not in entities:
            fail("MISSING_TARGET", "Evidence target is missing")
        part = require_revision(parts, link["part"], "evidence part")
        precise = any(locator["kind"] in policy["preciseLocators"] and (locator["kind"] != "json_pointer" or locator["pointer"] != "") for locator in part["locators"])
        if link["purpose"] in policy["exactPurposes"] and not precise:
            fail("LOCATOR_NOT_QUALIFIED", "Locator does not identify a qualified geometry/level part")
        inherited = link["inheritance"]
        if inherited["kind"] == "inherited":
            if link["purpose"] not in policy["inheritedPurposes"]:
                fail("INHERITANCE_PURPOSE", "Inherited documents are not exact unit geometry")
            pinned = inherited["parentLink"]
            if (not is_current or link["state"] == "unlinked") and version_key(pinned) in history:
                parent = history[version_key(pinned)]
            else:
                parent = require_revision(links, pinned, "parent evidence link")
            if parent["inheritance"]["kind"] != "direct" or (link["state"] == "active" and parent["state"] != "active") or version_key(parent["part"]) != version_key(link["part"]):
                fail("INHERITANCE_PARENT", "Inherited evidence needs a current direct association")
            cursor = ref_key(link["target"])
            visited = {cursor}
            for relation_id in inherited["via"]:
                edge = relations.get(relation_id)
                if edge is None or edge["kind"] not in policy["inheritanceRelations"] or ref_key(edge["from"]) != cursor:
                    fail("INHERITANCE_PATH", "Evidence does not follow a qualified containment path")
                cursor = ref_key(edge["to"])
                if cursor in visited:
                    fail("INHERITANCE_PATH", "Evidence inheritance repeats an identity")
                visited.add(cursor)
            if cursor != ref_key(parent["target"]):
                fail("INHERITANCE_PATH", "Evidence path does not reach its parent association")
        if is_current and link["state"] == "active":
            edge_keys.append((ref_key(link["target"]), version_key(link["part"]), link["purpose"], structural_key(inherited)))
    unique(edge_keys, "active evidence link")
    return catalog
